src/func.h

```diff
@@ -65,12 +65,65 @@ class TemplateParms : public Traceable {
     std::vector<const TemplateTypeParmType *> parms;
 };
 
+// Represents a single argument in a template instantiation. This can either be a type
+// or a non-type (constant expression).
+class TemplateArg {
+  public:
+    enum class ArgType { Type, /*To be extended with NoneType*/ };
+
+  private:
+    ArgType argType;
+    union {
+        const Type *type;
+    };
+    SourcePos pos;
+
+  public:
+    TemplateArg(const Type *t, SourcePos pos);
+
+    // Returns ISPC type of the argument.
+    const Type *GetAsType() const;
+    // Returns the source position associated with this template argument.
+    SourcePos GetPos() const;
+    // Produces a string representation of the argument.
+    std::string GetString() const;
+    // Returns `true` if this argument is a Type.
+    bool IsType() const;
+    // Compares two `TemplateArg` instances for equality.
+    bool IsEqual(const TemplateArg &other) const;
+    // Mangles the stored type to a string representation.
+    std::string Mangle() const;
+    // Transforms the stored type to its varying equivalent.
+    void SetAsVaryingType();
+};
+
+// Represents a list of template arguments. This is used
+// to store all arguments provided to a template, preserving their order and types.
 class TemplateArgs {
```

Comment:
Do we really need a class here? Why `typedef std::vector<TemplateArg> TemplateArgs` is not enough?

Reply:
I created a separate class to maintain consistency with the `TemplateParams` class. However, I agree that it doesn't introduce any functionality beyond that of a basic `std::vector`. Additionally, in the forthcoming PR, I plan to introduce a `TemplateParam` class that will encapsulate the logic currently implemented in the `TemplateParams::IsEqual` method, so the `TemplateParams` class will become redundant as well. I am considering transitioning both to `std::vector`.

Reply:
Ok

Reply:
Replaced `TemplateArgs` class with typedef.